File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    public readonly int speed;
    public readonly int batteryDrain;
    private int distanceDriven = 0;
    private int batteryLeft = 100;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        return batteryLeft < batteryDrain;
    }

    public int DistanceDriven()
    {
        return distanceDriven;
    }

    public void Drive()
    {
        if (!BatteryDrained())
        {
            distanceDriven += speed;
            batteryLeft -= batteryDrain;
        }
    }

    public static RemoteControlCar Nitro()
    {
        return new(speed: 50, batteryDrain: 4);
    }
}

class RaceTrack
{
    private int distance;
    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        return (car.batteryDrain * distance/car.speed) <= 100;
    }
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    public final int speed;
    public final int batteryDrain;

    private int batteryLeft = 100;
    private int distanceDriven = 0;

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryLeft < batteryDrain;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (!batteryDrained()){
            distanceDriven += speed;
            batteryLeft -= batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private final int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return (car.batteryDrain * distance/car.speed) <= 100;
    }
}
